### ISSUE TITLE
"Make member readonly" is offered even in cases where it silently changes behavior

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#69645]
**Version Used**: Visual Studio Version 17.8.0 Preview 1.0

**Steps to Reproduce**:

In the following code, [IDE0251](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0251) is issued suggesting to make the `ToString` method readonly. However doing so can change the behavior of the program since it will then induce a copy of `Item`, and in case `T` is a struct whose `ToString` method mutates it, the mutation will no longer be visible. I think the analyzer shouldn't trigger in this case.
```c#
struct ValueTuple<T>
{
    public T Item;

    public override string? ToString()
    {
        return Item.ToString();
    }
}
```
If I change `T` to be a custom struct that overrides `ToString`, the analyzer will no longer trigger as expected.
```c#
struct ValueTuple<T>
{
    public S Item;

    public override string? ToString()
    {
        return Item.ToString();
    }
}

struct S
{
    public override string? ToString()
    {
        return base.ToString();
    }
}
```

**Diagnostic Id**: [IDE0251](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0251)

**Expected Behavior**:
No diagnostic is issued.

**Actual Behavior**:

**Comment by CyrusNajmabadi** [comment]
I feel like this is a dupe, but i can't find it.  Generally, we've done with this being acceptable as the number of times that mutation is expected/desirable is far less than the number of times when it is not.  As such, better to allow this, but have users disable in the rare case, versus pessimistically disallow in a lot of valid cases.

**Comment by Neme12** [comment]
But then this analyzer is unusable to use in places like dotnet/runtime :/

**Comment by Neme12** [comment]
It looks like there is already some logic there to detect this though, because in case the member is a struct and it's a call to a non-readonly method, the analyzer won't fire. If this is deemed acceptable by default and there should be a way to disable that, maybe it should fire by default in the case the member is a struct as well? That's the same issue, just non-generic.

**Comment by CyrusNajmabadi** [comment]
>But then this analyzer is unusable to use in places like dotnet/runtime :/

C'est la vie.

**Comment by CyrusNajmabadi** [comment]
> It looks like there is already some logic there to detect this though, because in case the member is a struct and it's a call to a non-readonly method, the analyzer won't fire.

Correct.  If the type is known to be a mutable struct, then we dont' offer.  The question is what to do with generics.  The position we took is that we view it as too onerous to assume the worst with generics, and that we would prefer to optimistically assume types were not doing strange things that depended on this in generic cases, versus pessmistically assuming they might be.
